opta/linker.py

```diff
@@ -0,0 +1,416 @@
+from collections.abc import MutableSequence
+from dataclasses import dataclass, field
+from typing import Any, Dict, FrozenSet, Iterable, List, Set
+
+from opta.layer2 import Layer
+from opta.link import Link
+from opta.link_spec import InputLinkSpec, OutputLinkSpec
+from opta.module2 import Module
+from opta.module_spec import ModuleSpec
+from opta.utils.ref import (
+    InterpolatedReference,
+    Reference,
+    SimpleInterpolatedReference,
+    get_all_references,
+    parse_ref_string,
+)
+from opta.utils.visit import Visitor, fill_missing_list_or_dict
+
+_ModuleMap = Dict[str, Module]
+
+
+@dataclass
+class LinkResult:
+    execution_order: List[FrozenSet[str]] = field(default_factory=list)
+    interpolations: Dict[str, Dict[Reference, InterpolatedReference]] = field(
+        default_factory=dict
+    )
+
+
+class Linker:
+    def __init__(self, module_specs: Iterable[ModuleSpec]):
+        self._module_specs = {spec.name: spec for spec in module_specs}
+
+    def process(self, layer: Layer) -> LinkResult:
+        result = LinkResult()
+        module_map = self._module_map(layer.modules)
+
+        result.interpolations = self._pass_interpolation(module_map)
+
+        # Pass: TODO Run validation of input, but ignore defaults or required; assume refs are valid
+        # Pass: TODO Make sure that there are no interpolations in the link params
+
+        # Pass: TODO Run validation of input, but ignore defaults or required; assume refs are valid
+        # Pass: TODO Make sure that there are no interpolations in the link params
+
+        self._pass_automatic_links(module_map)
+        self._pass_add_links_from_interpolation(module_map, result.interpolations)
+
+        # Pass: connect outputs to links; TODO: This might not actually be needed? Can happen during input phase
+        # Pass: verify that we don't have any cyclical links
+
+        self._pass_connect_inputs(module_map)
+        # Pass: Validate input
+        # Pass: Run validation of input, but assume refs are valid
+        # Pass: build execution order
+
+        order = self._pass_build_execution_order(module_map)
+        result.execution_order = order
```

Comment:
unit tests checking the execution order for different scenarios would help understand this code